File: backend/app/schemas/report.py
from datetime import datetime
from enum import Enum
from typing import Optional
from pydantic import BaseModel, Field

class ReportStatus(str, Enum):
    PENDING = "pending"
    INVESTIGATING = "investigating"
    RESOLVED = "resolved"
    DISMISSED = "dismissed"

class ReportSeverity(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"
    UNKNOWN = "unknown"

class ReportCategory(str, Enum):
    TRACK = "TRACK"
    SIGNAL = "SIGNAL"
    TRACTION_OHE = "TRACTION_OHE"
    OTHER = "OTHER"

class ReportBase(BaseModel):
    category: ReportCategory = Field(ReportCategory.OTHER, description="Category of the railway issue")
    photo_url: Optional[str] = Field(None, description="URL or reference key to the stored photo")
    latitude: float = Field(..., ge=-90.0, le=90.0, description="GPS latitude coordinates (-90 to 90)")
    longitude: float = Field(..., ge=-180.0, le=180.0, description="GPS longitude coordinates (-180 to 180)")
    description: str = Field(..., min_length=1, description="Detailed description of the report/issue")
    timestamp: Optional[datetime] = Field(None, description="Timestamp of when the issue was reported/observed")
    reporter_id: Optional[str] = Field(None, description="Unique identifier of the reporting worker")

class ReportCreate(ReportBase):
    pass

class ReportUpdate(BaseModel):
    status: Optional[ReportStatus] = Field(None, description="Update status of the report")
    severity: Optional[ReportSeverity] = Field(None, description="Update severity level of the report")
    category: Optional[ReportCategory] = Field(None, description="Update category of the report")
    description: Optional[str] = Field(None, min_length=1, description="Update report description")
    latitude: Optional[float] = Field(None, ge=-90.0, le=90.0, description="Update GPS latitude")
    longitude: Optional[float] = Field(None, ge=-180.0, le=180.0, description="Update GPS longitude")
    photo_url: Optional[str] = Field(None, description="Update photo URL")

class ReportResponse(ReportBase):
    id: str = Field(..., description="Unique UUID identifier for the report")
    serverId: str = Field(..., description="Alias of id for mobile app client sync compatibility")
    status: ReportStatus = Field(..., description="Current status of the report")
    severity: ReportSeverity = Field(..., description="Assessed severity level")
    created_at: datetime = Field(..., description="Record creation timestamp")
    updated_at: datetime = Field(..., description="Record last updated timestamp")

    class Config:
        from_attributes = True

File: flutter_field_app/lib/models/report.dart
import 'package:flutter/material.dart';
import 'package:railsetu_field_app/core/theme/app_theme.dart';

/// Report category — aligned with backend ReportCategory enum.
enum ReportCategory {
  track('Track', Icons.train),
  signal('Signal', Icons.traffic),
  electrical('Electrical', Icons.bolt),
  station('Station', Icons.apartment),
  bridge('Bridge', Icons.landscape),
  safety('Safety', Icons.health_and_safety),
  other('Other', Icons.report_problem);

  final String label;
  final IconData icon;
  const ReportCategory(this.label, this.icon);
}

/// Report severity — aligned with backend ReportSeverity enum.
enum ReportSeverity {
  low('Low', AppColors.severityLow),
  medium('Medium', AppColors.severityMedium),
  high('High', AppColors.severityHigh),
  critical('Critical', AppColors.severityCritical);

  final String label;
  final Color color;
  const ReportSeverity(this.label, this.color);
}

/// Report status — aligned with backend ReportStatus enum.
enum ReportStatus {
  pending('Pending'),
  investigating('Under Review'),
  resolved('Resolved'),
  dismissed('Dismissed');

  final String label;
  const ReportStatus(this.label);
}

/// Core report model.
/// Mirrors the backend's ReportResponse schema fields.
class Report {
  final String id;
  final String title;
  final ReportCategory category;
  final ReportSeverity severity;
  final ReportStatus status;
  final String description;
  final double latitude;
  final double longitude;
  final String? photoUrl;
  final String? reporterId;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Report({
    required this.id,
    required this.title,
    required this.category,
    required this.severity,
    required this.status,
    required this.description,
    required this.latitude,
    required this.longitude,
    this.photoUrl,
    this.reporterId,
    required this.createdAt,
    required this.updatedAt,
  });

  Report copyWith({
    String? id,
    String? title,
    ReportCategory? category,
    ReportSeverity? severity,
    ReportStatus? status,
    String? description,
    double? latitude,
    double? longitude,
    String? photoUrl,
    String? reporterId,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Report(
      id: id ?? this.id,
      title: title ?? this.title,
      category: category ?? this.category,
      severity: severity ?? this.severity,
      status: status ?? this.status,
      description: description ?? this.description,
      latitude: latitude ?? this.latitude,
      longitude: longitude ?? this.longitude,
      photoUrl: photoUrl ?? this.photoUrl,
      reporterId: reporterId ?? this.reporterId,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  /// Convert to JSON map for creating a report — matching backend ReportCreate schema.
  Map<String, dynamic> toCreateJson() {
    return {
      'category': category.name.toUpperCase(),
      'severity': severity.name,
      'photo_url': photoUrl,
      'latitude': latitude,
      'longitude': longitude,
      'description': description,
      'timestamp': createdAt.toUtc().toIso8601String(),
      'reporter_id': reporterId,
    };
  }

  /// Convert to JSON map — ready for future API integration.
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'category': category.name.toUpperCase(),
      'severity': severity.name,
      'status': status.name,
      'description': description,
      'latitude': latitude,
      'longitude': longitude,
      'photo_url': photoUrl,
      'reporter_id': reporterId,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  /// Parse from JSON map — ready for future API integration.
  factory Report.fromJson(Map<String, dynamic> json) {
    return Report(
      id: json['id'] as String? ?? '',
      title: json['title'] as String? ?? json['category'] as String? ?? 'Report',
      category: ReportCategory.values.firstWhere(
        (e) => e.name.toUpperCase() == (json['category'] as String?)?.toUpperCase(),
        orElse: () => ReportCategory.other,
      ),
      severity: ReportSeverity.values.firstWhere(
        (e) => e.name.toLowerCase() == (json['severity'] as String?)?.toLowerCase(),
        orElse: () => ReportSeverity.low,
      ),
      status: ReportStatus.values.firstWhere(
        (e) => e.name.toLowerCase() == (json['status'] as String?)?.toLowerCase(),
        orElse: () => ReportStatus.pending,
      ),
      description: json['description'] as String? ?? '',
      latitude: (json['latitude'] as num?)?.toDouble() ?? 0.0,
      longitude: (json['longitude'] as num?)?.toDouble() ?? 0.0,
      photoUrl: json['photo_url'] as String?,
      reporterId: json['reporter_id'] as String?,
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at'] as String) : DateTime.now(),
      updatedAt: json['updated_at'] != null ? DateTime.parse(json['updated_at'] as String) : DateTime.now(),
    );
  }
}
